sdk/include/debug.hh

```diff
@@ -741,4 +741,59 @@ namespace
 		Assert(auto, const char *, Ts &&...) -> Assert<Ts...>;
 	};
 
+	/**
+	 * Check the stack usage of a function.  If `Enabled` is true, this will
+	 * print a message if the stack usage is larger than `Expected`.  One
+	 * message will be printed for each invocation of the function that exceeds
+	 * the previous maximum stack usage.
+	 *
+	 * An instance of this should be created as a local variable on entry to a
+	 * function.  The destructor (which prints the message) will then run after
+	 * any other destructors, ensuring the most accurate stack usage
+	 * measurement.
+	 *
+	 * Unfortunately, default arguments for templates are not evaluated in the
+	 * enclosing scope and so `__builtin_FUNCTION()` cannot be used here.
+	 * Instead, we must pass this explicitly, typically as something like:
+	 *
+	 * ```c++
+	 * StackUsageCheck<DebugFlag, 128, __PRETTY_FUNCTION__> stackCheck;
+	 * ```
+	 */
+	template<bool Enabled, size_t Expected, DebugContext Fn>
+	class StackUsageCheck
+	{
+		/**
+		 * The expected maximum.  This class is templated on the function name
+		 * and so there is one copy of this per function.
+		 */
+		static inline size_t stackUsage = Expected;
+
+		public:
+		/**
+		 * Default constructor, does nothing.
+		 */
+		StackUsageCheck() = default;
+
+		/**
+		 * Destructor, runs at the end of the function to print the message.
+		 */
+		__always_inline ~StackUsageCheck()
+		{
+			if constexpr (Enabled)
+			{
+				ptraddr_t lowest = stack_lowest_used_address();
+				ptraddr_t highest =
+				  CHERI::Capability{__builtin_cheri_stack_get()}.top();
+				size_t used = highest - lowest;
+				if (used > stackUsage)
+				{
+					stackUsage = used;
+					ConditionalDebug<Enabled, Fn>::log("Stack used: {} bytes",
```

Comment:
```suggestion
					ConditionalDebug<Enabled, Fn>::log("Stack use exceeded 
					{}: {} bytes", Expected,
```

Reply:
Why? This is printing the amount used.

Reply:
Just thought it might be useful to have both the expected and actual stack usage.